src/xo-app/backup/new/index.js

```diff
@@ -274,6 +287,80 @@ const BACKUP_METHOD_TO_INFO = {
 
 // ===================================================================
 
+const SAMPLE_SIZE_OF_MATCHING_VMS = 3
+
+@connectStore({
+  vms: createGetObjectsOfType('VM'),
+})
+class SmartBackupPreview extends Component {
+  static propTypes = {
+    pattern: PropTypes.object.isRequired,
+  }
+
+  _createPredicate = createSelector(
+    pattern => pickBy(pattern, val => val != null),
+    createPredicate
+  )
+
+  _getMatchingVms = createSelector(
+    () => this.props.pattern,
+    () => this.props.vms,
+    (pattern, vms) => filter(vms, this._createPredicate(pattern))
+  )
+
+  render () {
+    const matchingVms = this._getMatchingVms()
+    const nMatchingVms = matchingVms.length
+    const sampleOfMatchingVms = sampleSize(
+      matchingVms,
+      SAMPLE_SIZE_OF_MATCHING_VMS
+    )
+    const queryString = constructQueryString(this.props.pattern)
```

Comment:
The query is still constructed in the render, and it probably does unnecessary work: use a selector.